components/script_bindings/str.rs

```diff
@@ -293,6 +298,62 @@ impl DOMString {
     }
 }
 
+/// Because this converts to a DOMString it become UTF-8 encoded which is closer to
+/// the spec definition of <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-json-bytes>
+/// but we generally do not operate on anything that is truly a WTF-16 string.
+///
+/// <https://infra.spec.whatwg.org/#serialize-a-javascript-value-to-a-json-string>
+pub fn serialize_jsval_to_json_utf8(
+    cx: SafeJSContext,
+    jsval: HandleValue,
+) -> Result<DOMString, Error> {
+    #[repr(C)]
+    struct VoidableString {
+        is_void: bool,
+        data: String,
+    }
+
+    let mut out_str = VoidableString {
+        is_void: true,
+        data: String::new(),
+    };
+
+    #[allow(unsafe_code)]
+    unsafe extern "C" fn stringified(
+        string: *const u16,
+        len: u32,
+        data: *mut std::ffi::c_void,
+    ) -> bool {
+        let s = data as *mut VoidableString;
+        let string_chars = slice::from_raw_parts(string, len as usize);
+        (*s).is_void = false;
+        (*s).data.push_str(&String::from_utf16_lossy(string_chars));
+        true
+    }
+    rooted!(in(*cx) let mut data = jsval.get());
+
+    unsafe {
+        let stringify_result = ToJSON(
+            *cx,
+            data.handle(),
+            HandleObject::null(),
+            HandleValue::null(),
+            Some(stringified),
+            &mut out_str as *mut VoidableString as *mut _,
+        );
+        if !stringify_result {
+            return Err(Error::JSFailed);
+        }
+    }
+
+    // ToJSON will not call the callback if the data cannot be serialized
```

Comment:
// 2. If result is undefined, then throw a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).
// Note: ToJSON will not call the callback if the data cannot be serialized.